dev/tools/lib/archive_publisher.dart

```diff
@@ -0,0 +1,175 @@
+
+  @override
+  String toString() {
+    String output = 'ArchivePublisherException';
+    if (message != null) {
+      output += ': $message';
+    }
+    final String stderr = result?.stderr ?? '';
+    if (stderr.isNotEmpty) {
+      output += ':\n$result.stderr';
+    }
+    return output;
+  }
+}
+
+/// Publishes the archive created for a particular version and git hash to
+/// the releases directory on cloud storage, and updates the metadata for
+/// releases.
+class ArchivePublisher {
+  ArchivePublisher(
+    this.revision,
+    this.version,
+    this.channel, {
+    this.processManager = const LocalProcessManager(),
+    this.tempDir,
+  });
+
+  /// A git hash describing the revision to publish. It should be the complete
+  /// hash, not just a prefix.
+  final String revision;
+
+  /// A version number for the release (e.g. "1.2.3").
+  final String version;
+
+  /// The channel to publish to. Can be either "dev" or "beta".
+  // TODO(gspencer): support beta: it is currently unimplemented.
+  final String channel;
+
+  /// The process manager to use for invoking commands. Typically only
+  /// used for testing purposes.
+  final ProcessManager processManager;
+
+  /// The temporary directory used for this publisher. If not set, one will
+  /// be created, used, and then removed automatically. If set, it will not be
+  /// deleted when done: that is left to the caller. Typically used by tests.
+  Directory tempDir;
+
+  static String gsBase = 'gs://flutter_infra';
+  static String releaseFolder = '/releases';
+  static String baseUrl = 'https://storage.googleapis.com/flutter_infra';
+  static String archivePrefix = 'flutter_';
+  static String releaseNotesPrefix = 'release_notes_';
+
+  final String metadataGsPath = '$gsBase$releaseFolder/releases.json';
+
+  /// Publishes the archive for the given constructor parameters.
+  bool publishArchive() {
+    assert(channel == 'dev', 'Channel must be dev (beta not yet supported)');
+    // Check for access early so that we don't try to publish things if the
+    // user doesn't have access to the metadata file.
+    _checkForGSUtilAccess();
+    final List<String> platforms = <String>['linux', 'mac', 'win'];
+    final Map<String, String> metadata = <String, String>{};
+    for (String platform in platforms) {
+      final String src = _builtArchivePath(platform);
+      final String dest = _destinationArchivePath(platform);
+      final String srcGsPath = '$gsBase$src';
+      final String destGsPath = '$gsBase$releaseFolder$dest';
+      _cloudCopy(srcGsPath, destGsPath);
+      metadata['${platform}_archive'] = '$channel/$platform$dest';
+    }
+    metadata['release_date'] = new DateTime.now().toUtc().toIso8601String();
+    metadata['version'] = version;
+    _updateMetadata(metadata);
+    return true;
+  }
+
+  /// Checks to make sure the user has access to the Google Storage bucket
+  /// required to publish. Will print an error and return false if not.
```

Comment:
looks like this throws instead of returning false as documented?

Reply:
Fixed.